newton/_src/solvers/vbd/rigid_vbd.py

```diff
@@ -0,0 +1,2672 @@
+    x0 = (y0 - L10 * x1 - L20 * x2) / L00
+
+    return wp.vec3(x0, x1, x2)
+
+
+@wp.func
+def rotation_vector_from_delta(R_delta: wp.mat33) -> wp.vec3:
+    """Rotation vector (axis-angle) from relative rotation matrix using v-branch log map.
+
+    Numerically safe small-angle and near-pi fallbacks are used.
+
+    Args:
+        R_delta: Relative rotation matrix (any frame).
+
+    Returns:
+        wp.vec3: Rotation vector theta*axis in the same frame as R_delta.
+    """
+    # vee(R - R^T) without materializing the full skew matrix
+    vx = R_delta[2, 1] - R_delta[1, 2]
+    vy = R_delta[0, 2] - R_delta[2, 0]
+    vz = R_delta[1, 0] - R_delta[0, 1]
+    v2 = vx * vx + vy * vy + vz * vz
+
+    # theta from trace and |vee|: theta = atan2( ||vee||/2, (trace(R)-1)/2 )
+    s = 0.5 * wp.sqrt(v2)
+    tr = R_delta[0, 0] + R_delta[1, 1] + R_delta[2, 2]
+    c = 0.5 * (tr - 1.0)
+    c = wp.clamp(c, -1.0, 1.0)
+    theta = wp.atan2(s, c)
+
+    if v2 > _SMALL_ANGLE_EPS2:
+        inv_norm_v = 1.0 / wp.sqrt(v2)
+        scale = theta * inv_norm_v
+        return wp.vec3(vx * scale, vy * scale, vz * scale)
+
+    # Small-angle or near-pi fallbacks
+    if c > 0.0:
+        # Small angle: theta * axis ~ 0.5 * vee(R - R^T)
+        return wp.vec3(0.5 * vx, 0.5 * vy, 0.5 * vz)
+
+    # Near-pi: recover axis from diagonals, then return theta * axis
+    # Note: This branch is rare in typical simulation (requires approximately 180-degree rotation between consecutive frames)
+    ax2 = 0.5 * (R_delta[0, 0] + 1.0)
+    ay2 = 0.5 * (R_delta[1, 1] + 1.0)
+    az2 = 0.5 * (R_delta[2, 2] + 1.0)
+    if ax2 >= ay2 and ax2 >= az2:
+        ax = wp.sqrt(wp.max(ax2, 0.0))
+        ay = (R_delta[0, 1] + R_delta[1, 0]) * 0.25 / (ax + 1.0e-12)
+        az = (R_delta[0, 2] + R_delta[2, 0]) * 0.25 / (ax + 1.0e-12)
+        axis = wp.normalize(wp.vec3(ax, ay, az))
+    elif ay2 >= az2:
+        ay = wp.sqrt(wp.max(ay2, 0.0))
+        ax = (R_delta[0, 1] + R_delta[1, 0]) * 0.25 / (ay + 1.0e-12)
+        az = (R_delta[1, 2] + R_delta[2, 1]) * 0.25 / (ay + 1.0e-12)
+        axis = wp.normalize(wp.vec3(ax, ay, az))
+    else:
+        az = wp.sqrt(wp.max(az2, 0.0))
+        ax = (R_delta[0, 2] + R_delta[2, 0]) * 0.25 / (az + 1.0e-12)
+        ay = (R_delta[1, 2] + R_delta[2, 1]) * 0.25 / (az + 1.0e-12)
+        axis = wp.normalize(wp.vec3(ax, ay, az))
+    return axis * theta
+
+
+@wp.func
+def cable_get_kappa(q_wp: wp.quat, q_wc: wp.quat, q_wp_rest: wp.quat, q_wc_rest: wp.quat) -> wp.vec3:
+    """Compute cable bending curvature vector kappa in the parent frame.
+
+    Kappa is the rotation vector (theta*axis) from the rest-aligned relative rotation.
+
+    Args:
+        q_wp: Parent orientation (world).
+        q_wc: Child orientation (world).
+        q_wp_rest: Parent rest orientation (world).
+        q_wc_rest: Child rest orientation (world).
+
+    Returns:
+        wp.vec3: Curvature vector kappa in parent frame (rotation vector form).
+    """
```

Comment:
Is it necessary/useful to go through matrices here? Since the numbers of mults per matrix is very low I believe staying with quats should give better perf